src/test/resources/com/jpinpoint/perf/lang/kotlin/ruleset/common/xml/AvoidRecreatingDateTimeFormatter.xml

```diff
@@ -0,0 +1,128 @@
+
+    fun testNoViolation_JPCC_16(dateFormat: String) {
+        val dateTimeFormatter = getDateTimeFormatterFromCacheViolation(dateFormat)
+    }
+
+    private fun getDateTimeFormatterFromCacheViolation(dateFormat: String): DateTimeFormatter {
+        return DateTimeFormatterBuilder().toFormatter() // bad
+    }
+
+    fun testNoViolation_2(dateFormat: String?) {
+        val dateTimeFormatter: DateTimeFormatter =
+            DateTimeFormat.forPattern(dateFormat) // good: possible different dataFormat params
+    }
+
+    fun testViolationThreeten() {
+        val ftor: DateTimeFormatter = org.threeten.bp.format.DateTimeFormatter.ofPattern("yyMMdd") // bad
+    }
+
+    fun testViolationJava8() {
+        val ftor = DateTimeFormatter.ofPattern("yyMMdd") // bad
+    }
+}
+]]></code>
+    </test-code>
+    <test-code>
+        <description>Avoid recreation of DateTimeFormatter object. But allow in constructor if member variable is val.</description>
+        <expected-problems>3</expected-problems>
+        <expected-linenumbers>13,14,15</expected-linenumbers>
+        <code><![CDATA[
+import org.joda.time.format.DateTimeFormat
+import org.joda.time.format.DateTimeFormatter
+import org.joda.time.format.DateTimeFormatterBuilder
+
+class Foo(pattern: String?) {
+
+    companion object {
+        private val ok2 = DateTimeFormat.forPattern("YY-MM-hh")
+    }
+
+    private val ok: DateTimeFormatter = DateTimeFormat.forPattern(pattern)
+    private val notok: DateTimeFormatter =
+        DateTimeFormat.forPattern("YY-MM-hh") // bad
+    private val notok2: DateTimeFormatter = DateTimeFormat.forPattern("YY-MM-hh") // bad (should be in companion object)
+    private var notok3 = DateTimeFormatterBuilder().appendPattern(pattern).toFormatter() // bad (should be val)
+}
+     ]]></code>
+    </test-code>
+    <test-code>
+        <description>Avoid recreation of DateTimeFormatter object. But allow if there is a parameter from method call involved.</description>
+        <expected-problems>3</expected-problems>
+        <expected-linenumbers>8,9,10</expected-linenumbers>
+        <code><![CDATA[
+import org.joda.time.format.DateTimeFormat
+import java.time.ZoneId
+
+object Foo {
+    private const val DEFAULT_DATETIME_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"
+    fun format(milliseconds: Long, myFormat: String?, myTimeZone: String?): String {
+        val formatter1 = DateTimeFormat.forPattern(myFormat) // good: has variable parameter
+        val formatter2 = DateTimeFormat.forPattern(DEFAULT_DATETIME_FORMAT) // bad
+        val formatter3 = DateTimeFormat.forPattern(DEFAULT_DATETIME_FORMAT).withZoneUTC() // bad
+        val formatter4 = DateTimeFormat.forPattern(DEFAULT_DATETIME_FORMAT).withZone(ZoneId.of("UTC")) // bad
+        val formatter5 =
+            DateTimeFormat.forPattern(DEFAULT_DATETIME_FORMAT).withZone(myTimeZone) // good: has variable parameter
+        return formatter1.print(milliseconds)
+    }
+}
+     ]]></code>
+    </test-code>
```

Comment:
Can you add a separate test-code with just java.time import? (no joda.time) 
and one without joda.time and without java.time and a custom DateTimeFormatter so without violation?

Reply:
These are already tests with only java.time and only joda.time.

Added custom DateTimeFormatter no violation test. And also one with just java.time.

Reply:
expected one unit test to fail without the fix